# Правило сложения вероятностей

В данном разделе будет рассмотренно сложение вероятностей событий

Пример:
- Событие A - нечетное число при броске игральной кости
- Событие B - число больше 2 при броске игральной кости

Задача:
- Расчитать вероятность возникновения каждого события по отдельности
- Расчитать вероятность возникновения двух событий одновременно

## Решение

Для нахождения вероятности возникновения событий по отдельности используется формула объединения событий которая рассчитывается как

```
P(A or B) = P(A) + P(B) - P(A and B)
```

Другими словами вероятность возникновения каждого события по отдельности это

**Сумма вероятностей каждого события и вычитание пересечения событий**

Мы вычитаем пересечение событий так как оно дважды появляется в каждой событии по отдельности

## А что если два события были бы взаимоисключающими
(**Mutually Exclusive**)

В таком случае формула выглядела бы так

```
P(A or B) = P(A) + P(B)
```

В таком случае мы не отнимаем пересечение двух событий так как его нет

# Python

In [1]:
# Напишем функцию которая позволит нам выявить вероятность возникновения каждого события по отдельности
def prob_a_or_b(a, b, all_possible_outcomes):
    # a - 1 событие
    # b - 2 событие
    # Возможные значения
    # Вероятность события A и B рассичитывается как отношение кол-ва значений события к общему кол-ву возможных значений
    prob_a = len(a)/len(all_possible_outcomes)
    prob_b = len(b)/len(all_possible_outcomes)
    # Найдем пересечение значений множеств
    inter = a.intersection(b)
    # Расчитаем вероятность события в пересечении
    prob_inter = len(inter)/len(all_possible_outcomes)
    # Возвращаем значение вероятности возникновения событий по отдельности
    return prob_a + prob_b - prob_inter

In [2]:
# Первый эксперимент
# Бросаем игральную кость с 6-ю гранями
# Первое событие проверяет является ли значение четным
# Второе событие проверяет является ли значение не четным
evens = {2,4,6}
odds = {1,3,5}
all_possible_rolls = {1,2,3,4,5,6}

print(prob_a_or_b(evens, odds, all_possible_rolls))

1.0


## Результат первого эксперимента
Получили 100% вероятность того что произойдет хотя бы одно событие из двух

И правда при любом кол-ве бросков значение будет либо четным либо не четным

In [5]:
# Второй эксперимент
# Бросаем игральную кость
# Первое событие значение будет не четным
# Второе событие значение будет больше 2

odds = {1,3,5}
greater_than_two = {3,4,5,6}
all_possible_rolls = {1,2,3,4,5,6}

print(prob_a_or_b(odds, greater_than_two, all_possible_rolls))

0.8333333333333333


## Результат второго эксперимента
Получили вероятность возникновения хотя бы одного из событий равному 83%

In [6]:
# Третий эксперимент
# Вытягиваем карту
# Первое событие масть карты буби
# Второе событие лицевая сторона карты будет картинкой

diamond_cards = {'ace_diamond', '2_diamond', '3_diamond', '4_diamond', '5_diamond', '6_diamond', '7_diamond', '8_diamond', '9_diamond', '10_diamond', 'jack_diamond', 'queen_diamond', 'king_diamond'}
face_cards = {'jack_diamond', 'jack_spade', 'jack_heart', 'jack_club', 'queen_diamond', 'queen_spade', 'queen_heart', 'queen_club', 'king_diamond', 'king_spade', 'king_heart', 'king_club'}
all_possible_cards = {'ace_diamond', '2_diamond', '3_diamond', '4_diamond', '5_diamond', '6_diamond', '7_diamond', '8_diamond', '9_diamond', '10_diamond', 'jack_diamond', 'queen_diamond', 'king_diamond', 'ace_heart', '2_heart', '3_heart', '4_heart', '5_heart', '6_heart', '7_heart', '8_heart', '9_heart', '10_heart', 'jack_heart', 'queen_heart', 'king_heart', 'ace_spade', '2_spade', '3_spade', '4_spade', '5_spade', '6_spade', '7_spade', '8_spade', '9_spade', '10_spade', 'jack_spade', 'queen_spade', 'king_spade', 'ace_club', '2_club', '3_club', '4_club', '5_club', '6_club', '7_club', '8_club', '9_club', '10_club', 'jack_club', 'queen_club', 'king_club'}

print(prob_a_or_b(diamond_cards, face_cards, all_possible_cards))

0.4230769230769231


## Результат третьего эксперимента
В результате получили вероятность возникновения хотя бы одного из событий равному 42%